File: examples/grpc-service/src/test/java/example/armeria/grpc/HelloServiceTest.java
package example.armeria.grpc;

import static example.armeria.grpc.HelloServiceImpl.toMessage;
import static org.assertj.core.api.Assertions.assertThat;
import static org.awaitility.Awaitility.await;

import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;

import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;

import com.google.common.base.Stopwatch;
import com.google.common.util.concurrent.FutureCallback;
import com.google.common.util.concurrent.Futures;
import com.google.common.util.concurrent.ListenableFuture;
import com.google.common.util.concurrent.MoreExecutors;

import com.linecorp.armeria.client.Clients;
import com.linecorp.armeria.server.Server;

import example.armeria.grpc.Hello.HelloReply;
import example.armeria.grpc.Hello.HelloRequest;
import example.armeria.grpc.HelloServiceGrpc.HelloServiceBlockingStub;
import example.armeria.grpc.HelloServiceGrpc.HelloServiceFutureStub;
import example.armeria.grpc.HelloServiceGrpc.HelloServiceStub;
import io.grpc.stub.StreamObserver;

public class HelloServiceTest {

    private static Server server;
    private static HelloServiceStub helloService;

    @BeforeClass
    public static void beforeClass() throws Exception {
        server = Main.newServer(0, 0);
        server.start().join();
        helloService = Clients.newClient(uri(), HelloServiceStub.class);
    }

    @AfterClass
    public static void afterClass() {
        if (server != null) {
            server.stop().join();
        }
    }

    private static String uri() {
        return "gproto+http://127.0.0.1:" + server.activeLocalPort() + '/';
    }

    @Test
    public void getReply() {
        final HelloServiceBlockingStub helloService = Clients.newClient(uri(), HelloServiceBlockingStub.class);
        assertThat(helloService.hello(HelloRequest.newBuilder().setName("Armeria").build()).getMessage())
                .isEqualTo("Hello, Armeria!");
    }

    @Test
    public void getReplyWithDelay() {
        final HelloServiceFutureStub helloService = Clients.newClient(uri(), HelloServiceFutureStub.class);
        final ListenableFuture<HelloReply> future =
                helloService.lazyHello(HelloRequest.newBuilder().setName("Armeria").build());
        final AtomicBoolean completed = new AtomicBoolean();
        Futures.addCallback(future, new FutureCallback<HelloReply>() {
            @Override
            public void onSuccess(HelloReply result) {
                assertThat(result.getMessage()).isEqualTo("Hello, Armeria!");
                completed.set(true);
            }

            @Override
            public void onFailure(Throwable t) {
                // Should never reach here.
                throw new Error(t);
            }
        }, MoreExecutors.directExecutor());

        await().untilTrue(completed);
    }

    @Test
    public void getReplyFromServerSideBlockingCall() {
        final HelloServiceBlockingStub helloService = Clients.newClient(uri(), HelloServiceBlockingStub.class);
        final Stopwatch watch = Stopwatch.createStarted();
        assertThat(helloService.blockingHello(HelloRequest.newBuilder().setName("Armeria").build())
                               .getMessage()).isEqualTo("Hello, Armeria!");
        assertThat(watch.elapsed(TimeUnit.SECONDS)).isGreaterThanOrEqualTo(3);
    }

    @Test
    public void getLotsOfReplies() {
        final AtomicBoolean completed = new AtomicBoolean();
        helloService.lotsOfReplies(
                HelloRequest.newBuilder().setName("Armeria").build(),
                new StreamObserver<HelloReply>() {
                    private int sequence;

                    @Override
                    public void onNext(HelloReply value) {
                        assertThat(value.getMessage())
                                .isEqualTo("Hello, Armeria! (sequence: " + ++sequence + ')');
                    }

                    @Override
                    public void onError(Throwable t) {
                        // Should never reach here.
                        throw new Error(t);
                    }

                    @Override
                    public void onCompleted() {
                        assertThat(sequence).isEqualTo(5);
                        completed.set(true);
                    }
                });
        await().untilTrue(completed);
    }

    @Test
    public void blockForLotsOfReplies() throws Exception {
        final BlockingQueue<HelloReply> replies = new LinkedBlockingQueue<>();
        final AtomicBoolean completed = new AtomicBoolean();
        helloService.lotsOfReplies(
                HelloRequest.newBuilder().setName("Armeria").build(),
                new StreamObserver<HelloReply>() {

                    @Override
                    public void onNext(HelloReply value) {
                        replies.offer(value);
                    }

                    @Override
                    public void onError(Throwable t) {
                        // Should never reach here.
                        throw new Error(t);
                    }

                    @Override
                    public void onCompleted() {
                        completed.set(true);
                    }
                });
        int sequence = 0;
        while (!completed.get() || !replies.isEmpty()) {
            final HelloReply value = replies.poll(100, TimeUnit.MILLISECONDS);
            if (value == null) {
                // Timed out, try again.
                continue;
            }
            assertThat(value.getMessage())
                    .isEqualTo("Hello, Armeria! (sequence: " + ++sequence + ')');
        }
        assertThat(sequence).isEqualTo(5);
    }

    @Test
    public void sendLotsOfGreetings() {
        final String[] names = { "Armeria", "Grpc", "Streaming" };
        final AtomicBoolean completed = new AtomicBoolean();
        final StreamObserver<HelloRequest> request =
                helloService.lotsOfGreetings(new StreamObserver<HelloReply>() {
                    private boolean received;

                    @Override
                    public void onNext(HelloReply value) {
                        assertThat(received).isFalse();
                        received = true;
                        assertThat(value.getMessage())
                                .isEqualTo(toMessage(String.join(", ", names)));
                    }

                    @Override
                    public void onError(Throwable t) {
                        // Should never reach here.
                        throw new Error(t);
                    }

                    @Override
                    public void onCompleted() {
                        assertThat(received).isTrue();
                        completed.set(true);
                    }
                });

        for (String name : names) {
            request.onNext(HelloRequest.newBuilder().setName(name).build());
        }
        request.onCompleted();
        await().untilTrue(completed);
    }

    @Test
    public void bidirectionalHello() {
        final String[] names = { "Armeria", "Grpc", "Streaming" };
        final AtomicBoolean completed = new AtomicBoolean();
        final StreamObserver<HelloRequest> request =
                helloService.bidiHello(new StreamObserver<HelloReply>() {
                    private int received;

                    @Override
                    public void onNext(HelloReply value) {
                        assertThat(value.getMessage())
                                .isEqualTo(toMessage(names[received++]));
                    }

                    @Override
                    public void onError(Throwable t) {
                        // Should never reach here.
                        throw new Error(t);
                    }

                    @Override
                    public void onCompleted() {
                        assertThat(received).isEqualTo(names.length);
                        completed.set(true);
                    }
                });

        for (String name : names) {
            request.onNext(HelloRequest.newBuilder().setName(name).build());
        }
        request.onCompleted();
        await().untilTrue(completed);
    }
}

File: examples/grpc-kotlin/src/test/kotlin/example/armeria/grpc/kotlin/HelloServiceTest.kt
package example.armeria.grpc.kotlin

import com.google.common.base.Stopwatch
import com.linecorp.armeria.client.Clients
import com.linecorp.armeria.server.Server
import example.armeria.grpc.kotlin.Hello.HelloReply
import example.armeria.grpc.kotlin.Hello.HelloRequest
import example.armeria.grpc.kotlin.HelloServiceGrpcKt.HelloServiceCoroutineStub
import kotlinx.coroutines.flow.asFlow
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.runBlocking
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import java.util.concurrent.TimeUnit

class HelloServiceTest {

    @ParameterizedTest
    @MethodSource("uris")
    fun reply(uri: String) {
        runBlocking {
            val helloService = Clients.newClient(uri, HelloServiceCoroutineStub::class.java)
            assertThat(helloService.hello(HelloRequest.newBuilder().setName("Armeria").build()).message)
                .isEqualTo("Hello, Armeria!")
        }
    }

    @ParameterizedTest
    @MethodSource("uris")
    fun replyWithDelay(uri: String) {
        runBlocking {
            val helloService = Clients.newClient(uri, HelloServiceCoroutineStub::class.java)
            val reply: HelloReply = helloService.lazyHello(HelloRequest.newBuilder().setName("Armeria").build())
            assertThat(reply.message).isEqualTo("Hello, Armeria!")
        }
    }

    @ParameterizedTest
    @MethodSource("uris")
    fun replyFromServerSideBlockingCall(uri: String) {
        runBlocking {
            val helloService = Clients.newClient(uri, HelloServiceCoroutineStub::class.java)
            val watch = Stopwatch.createStarted()
            assertThat(helloService.blockingHello(HelloRequest.newBuilder().setName("Armeria").build()).message)
                .isEqualTo("Hello, Armeria!")
            assertThat(watch.elapsed(TimeUnit.SECONDS)).isGreaterThanOrEqualTo(3)
        }
    }

    @Test
    fun lotsOfReplies() {
        runBlocking {
            var sequence = 0
            helloService.lotsOfReplies(HelloRequest.newBuilder().setName("Armeria").build())
                .collect {
                    assertThat(it.message).isEqualTo("Hello, Armeria! (sequence: ${++sequence})")
                }
            assertThat(sequence).isEqualTo(5)
        }
    }

    @Test
    fun blockForLotsOfReplies() {
        runBlocking {
            val replies = ArrayList<HelloReply>()
            helloService.lotsOfReplies(HelloRequest.newBuilder().setName("Armeria").build())
                .collect { replies.add(it) }
            for ((sequence, reply) in replies.withIndex()) {
                assertThat(reply.message).isEqualTo("Hello, Armeria! (sequence: ${sequence + 1})")
            }
        }
    }

    @Test
    fun sendLotsOfGreetings() {
        runBlocking {
            val names = listOf("Armeria", "Grpc", "Streaming")
            val requests = names.map { HelloRequest.newBuilder().setName(it).build() }

            val reply: HelloReply = helloService.lotsOfGreetings(requests.asFlow())
            assertThat(reply.message).isEqualTo("Hello, ${names.joinToString()}!")
        }
    }

    @Test
    fun bidirectionalHello() {
        runBlocking {
            val names = listOf("Armeria", "Grpc", "Streaming")
            val requests = names.map { HelloRequest.newBuilder().setName(it).build() }
            val replies = helloService.bidiHello(requests.asFlow())

            var received = 0
            replies.collect {
                assertThat(it.message).isEqualTo("Hello, ${names[received++]}!")
            }
        }
    }

    companion object {

        private lateinit var server: Server
        private lateinit var blockingServer: Server
        private lateinit var helloService: HelloServiceCoroutineStub

        @BeforeAll
        @JvmStatic
        fun beforeClass() {
            server = Main.newServer(0, 0)
            server.start().join()

            blockingServer = Main.newServer(0, 0, true)
            blockingServer.start().join()
            helloService = Clients.newClient(protoUri(), HelloServiceCoroutineStub::class.java)
        }

        @AfterAll
        @JvmStatic
        fun afterClass() {
            server.stop().join()
            blockingServer.stop().join()
        }

        @JvmStatic
        fun uris() = listOf(protoUri(), jsonUri(), blockingProtoUri(), blockingJsonUri())
            .map { Arguments.of(it) }

        private fun protoUri(): String {
            return "gproto+http://127.0.0.1:" + server.activeLocalPort() + '/'
        }

        private fun jsonUri(): String {
            return "gjson+http://127.0.0.1:" + server.activeLocalPort() + '/'
        }

        private fun blockingProtoUri(): String {
            return "gproto+http://127.0.0.1:" + blockingServer.activeLocalPort() + '/'
        }

        private fun blockingJsonUri(): String {
            return "gjson+http://127.0.0.1:" + blockingServer.activeLocalPort() + '/'
        }
    }
}
